scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/web/DownloadDatasetController.java

```diff
@@ -0,0 +1,82 @@
+import edu.asu.diging.scriptoocloud.core.model.impl.Dataset;
+import edu.asu.diging.scriptoocloud.core.service.IDatasetService;
+import edu.asu.diging.scriptoocloud.core.service.IFileSystemService;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.security.access.prepost.PreAuthorize;
+import org.springframework.stereotype.Controller;
+import org.springframework.web.bind.annotation.*;
+
+import javax.servlet.http.HttpServletResponse;
+import java.io.*;
+import java.security.Principal;
+import java.util.Locale;
+
+@Controller
+public class DownloadDatasetController {
+
+    private final Logger logger = LoggerFactory.getLogger(getClass());
+
+    private final IDatasetService iDatasetService;
+    private final IFileSystemService iFileSystemService;
+
+    @Autowired
+    public DownloadDatasetController(IDatasetService iDatasetService, IFileSystemService iFileSystemService) {
+        this.iDatasetService = iDatasetService;
+        this.iFileSystemService = iFileSystemService;
+    }
+
+    @RequestMapping(value = "auth/datasets/{id}/download")
+    @PreAuthorize("hasPermission(#datasetId, 'Dataset', 'download')")
+    public void download(@PathVariable("id") Long datasetId,
+                         HttpServletResponse response,
+                         Principal principal) throws IOException, FileSystemStorageException {
+        String username = principal.getName();
+        String type = Dataset.class.getSimpleName().toLowerCase(Locale.ROOT);
+        Dataset dataset = (Dataset) iDatasetService.findById(datasetId);
+        if (dataset != null) {
+            String filename = type + "_" + datasetId + "_version_" + dataset.getVersion() + ".zip";
+            try {
+                iFileSystemService.createZipFile(username, type, datasetId.toString(),
+                        dataset.getVersion().toString());
+                File file = new File(iFileSystemService.createPath(username, type,
+                        datasetId.toString(),
+                        dataset.getVersion().toString()).toString() + File.separator + filename);
+                if (file.exists()) {
+                    response.setHeader("Content-Disposition", "attachment; filename=" + filename);
+                    response.setContentType("application/zip");
+                    OutputStream outputStream = response.getOutputStream();
+                    FileInputStream fileInputStream = new FileInputStream(file);
+                    byte[] buffer = new byte[4096];
+                    int bytesRead;
+                    while ((bytesRead = fileInputStream.read(buffer)) != -1) {
+                        outputStream.write(buffer, 0, bytesRead);
+                    }
+                    fileInputStream.close();
+                    outputStream.close();
+                    response.flushBuffer();
+
+                    iFileSystemService.deleteDirectoryOrFile(
```

Comment:
I would not delete the file. What if something goes wrong or the user wants to download the same file again? If it took 30min to create it, that would be really annoying to have to wait again. I would change the workflow so that:
- the user starts the zipping process
- this runs in the background
- there is another page that lists all previously created zips/snapshots
- once the zipping is done, the snapshot shows up on the list page (there could be a notification on the screen if you want to make it fancy ;)
- the user can download the zipped file from the page and also delete it from there if they want to

I would keep track of created zips by storing some metadata in the database (e.g. when it was created, what dataset, version, etc). Then it would be simple to list them all.